Multi Component Group Wrapper

vitvakatu opened the pull request:
### Pull Request Description

[ci no changelog needed]

This PR implements a new helper for the future Component Browser - `component_group::multi::Wrapper`. It propagates FRP events from multiple component groups and ensures that only a single component group is focused at all times.

See the updated component group demo scene (console logs shows propagated FRP events from all component groups):


https://user-images.githubusercontent.com/6566674/172359141-8ea6f1ba-e357-4c1b-852a-adb4d5207e03.mp4


- Fixed a `define_endpoints_2!` macro. FRP endpoints for `focus` events weren't connected properly.
- List View now uses an overlay shape to catch mouse events, it allows much easier implementation of `is_header_selected` in the component group.

### Important Notes

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run.sh ide dist` and `./run.sh ide watch`.

vitvakatu committed:
Implement multi component group wrapper

vitvakatu committed:
Merge branch 'develop' into multiview

vitvakatu committed:
self-reviw

vitvakatu committed:
self-review

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -431,7 +436,8 @@ impl component::Model for Model {
         entries.show_background_shadow(false);
         entries.set_background_corners_radius(0.0);
-        entries.hide_selection();
+        // FIXME(#182194574): Hide selection again once we have a proper selection box.
```
1. Add a real link to the issue pls, it is easier to navigate by cliking in intellij idea then.
2. Every fixme should describe shortly WHY it is not working right now. The link should be an additional explanation. Right now, after reading this fixme, I dont understand why we cant enable it right now and what does "proper selection box" mean.

vitvakatu replied:
I decided to revert this change, as with new styles default selection looks absolutely ugly.

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -462,6 +468,13 @@ impl Model {
         self.header.add_to_scene_layer(&layers.header_text);
     }
 
+    /// Whether the `point` (object-space coordinates) is inside the component group shape.
```
Minor thing: Comments should be written in active form. Use "Test whether ..." instead. Pls fix it here and in other comments.

wdanilo commented on app/gui/view/component-browser/component-group/src/multi.rs:
```diff
@@ -0,0 +1,212 @@
+// === Group ===
+// =============
+
+/// One of the possible component group types.
+#[derive(Debug, Clone, CloneRef)]
+#[allow(missing_docs)]
+pub enum Group {
+    Regular(View),
+    Wide(wide::View),
+}
```
These docs do not explain what it is. After reading it, I don't know what is "Regular" and "Wide". This should be described in such a way, that the reader understands without reading implementation of methods later in this file.

vitvakatu replied:
I renamed `Regular` -> `OneColumn`, but I'm unsure what else I should add. This is an enum with doc "One of the possible component group types." So by reading this doc, I understand that there are multiple component group types and they (or part of them) are listed in this enum. I can easily follow to the type I'm interested in. (`View` or `wide::View`)
I don't think it's wise to put all the information about component groups here. Let's chat about it if you disagree.

wdanilo commented on app/gui/view/component-browser/component-group/src/multi.rs:
```diff
@@ -0,0 +1,212 @@
+    Wide(wide::View),
+}
+
+impl From<View> for Group {
+    fn from(view: View) -> Self {
+        Group::Regular(view)
+    }
+}
+
+impl From<wide::View> for Group {
+    fn from(view: wide::View) -> Self {
+        Group::Wide(view)
+    }
```
you sohuld be able to derive it

vitvakatu replied:
Done

wdanilo commented on app/gui/view/component-browser/component-group/src/multi.rs:
```diff
@@ -0,0 +1,212 @@
+
+impl Wrapper {
+    /// Constructor.
+    pub fn new(scene: &Scene, network: &frp::Network, groups: impl Iterator<Item = Group>) -> Self {
+        let mouse = &scene.mouse.frp;
+        let groups = Rc::new(groups.collect::<Vec<_>>());
+        frp::extend! { network
+            mouse_in_group <- any(...);
+            selected_entry <- any(...);
+            suggestion_accepted <- any(...);
+            expression_accepted <- any(...);
+            is_header_selected <- any(...);
+            header_accepted <- any(...);
+            selection_position_target <- any(...);
+            focused <- any(...);
+        }
+
+        for (i, group) in groups.iter().enumerate() {
+            let id = GroupId::from(i);
+            frp::extend! { network
+                mouse_in <- map(&mouse.position,f!([scene,group](pos) {
+                    let pos_obj_space = group.to_object_space(&scene, *pos);
+                    group.is_inside(pos_obj_space)
+                }));
+                mouse_in_group <+ mouse_in.on_change().on_true().map(move |_| id);
+            }
+            match group {
+                Group::Regular(group) => {
+                    frp::extend! { network
+                        focused <+ group.focused.map(move |f| (id, *f));
```
One worry here – searcher can contain dozens/hundreds of groups. We should rather not create all of them upfront, we might want to create/remove them while scrolling on-demand instead. Otherwise, we will have performance issues here. Is this taken into account?

vitvakatu replied:
Good catch, I changed the design of the API slightly, now it has `add`/`remove` methods. The implementation was kinda reworked because of it.

wdanilo commented on app/gui/view/debug_scene/component-group/src/lib.rs:
```diff
@@ -162,6 +162,15 @@ fn create_component_group(
     component_group
 }
 
+fn create_wide_component_group(app: &Application) -> component_group::wide::View {
+    let component_group = app.new_view::<component_group::wide::View>();
+    component_group.set_width(450.0);
+    component_group.set_position_x(-200.0);
+    let background_color = color::Rgba(0.9, 0.91, 0.89, 1.0);
+    component_group.set_background_color(background_color);
+    component_group
+}
```
A lot of hardcoded values. Refactor to constants

vitvakatu replied:
I removed `background_color`. (it is no longer needed after merging with develop)
I don't think moving the other two literals (-200.0 and 450.0) into constants makes sense. They are purely example values and will probably last only until the next version of the demo scene.

wdanilo reviewed (changes requested):
1. Minor changes, but IMO important comments to be resolved, so I'm leaving request changes. Let's chat about them and merge it :)
2. Add animation to dimming pls - it hurts the eyes when it's changing immediately.

farmaazon committed:
Fix leaking component browser entry

vitvakatu commented:
Status report: I'm addressing review comments, implementing dynamic adding/removal of component groups takes more time than I thought. I'm drafting the PR in order not to confuse reviewers.

vitvakatu committed:
addressing review comments

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/multiview-182259906

vitvakatu committed:
formatting & clippy issues

vitvakatu commented:
The implementation was updated to support dynamic adding/removal of component groups. Also, I implemented dimming animation for component groups.

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -55,6 +56,7 @@ use ensogl_shadow as shadow;
 
 pub mod entry;
 pub mod icon;
+pub mod multi;
```
I'd name it 'set'. The word 'multi' is not a noun and doesnt express the intention as good as "set" imo. Do you agree?

wdanilo commented on app/gui/view/component-browser/component-group/src/multi.rs:
```diff
@@ -0,0 +1,255 @@
+// === Group ===
+// =============
+
+/// One of the possible component group types.
+#[derive(Debug, Clone, CloneRef, From)]
+#[allow(missing_docs)]
+pub enum Group {
+    OneColumn(View),
+    Wide(wide::View),
+}
```
I am almost sure I was writing in my previous review about it. These docs need to be expanded. What is "Wide"? I can't understand that when reading the code.

wdanilo commented on app/gui/view/component-browser/component-group/src/multi.rs:
```diff
@@ -0,0 +1,255 @@
+/// to connect two such structs. See [`PropagatedEvents`] docs.
+macro_rules! propagated_events {
+    (
+        $(#[$meta:meta])* struct $ident:ident {
```
Use `$(#meta:tt)` instead.

wdanilo commented on app/gui/view/component-browser/component-group/src/multi.rs:
```diff
@@ -0,0 +1,255 @@
+macro_rules! propagated_events {
+    (
+        $(#[$meta:meta])* struct $ident:ident {
+            $($endpoint:ident : $endpoint_type:ty),*$(,)?
```
add space between `*` and `$` to make it more clear.

wdanilo commented:
Btw, pls update the screencast! :)

vitvakatu committed:
rename multi -> set

vitvakatu committed:
minor fixes in macro

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/multiview-182259906

vitvakatu committed:
Fix invalid rename

vitvakatu committed:
Fix list-view demo

farmaazon commented:
Issues found:
1. Scenario: select header of some group; select some entry of the same group; after that I cannot select the header of the same group. The selecting is "enabled" only when selected header of another group. The selecting may be with mouse or keyboard
2. Not sure if relevant, but keyboard strokes which work in one-column component group (up/down arrows, enter) don't work in wide component group.

vitvakatu committed:
Fix spotted issues

vitvakatu committed:
Fix a few issues with selecting entries

vitvakatu committed:
Fix formatting

vitvakatu commented:
@farmaazon 
1. Fixed! The issue was in debug scene code actually. After fixing it, I noticed that header select and entry select events overlap each other in some cases, so I fixed that as well.
2. Nice catch, this is relevant! Fixed, now focus events correctly propagate to the wide component group as well.

farmaazon commented:
Bugs above are fixed, but there are new:
1. In wide component group, the entry selected signal is emitted with the slightest mouse movement, the behavior differs from the single-column view.
2. Not sure if related to this very PR, but when I move the mouse down from the last component group, I sometimes get the message about selecting an Entry which is apparently not visible.
https://user-images.githubusercontent.com/3919101/171188370-f4d4355b-df70-4100-9f46-06fdac50bdcd.mp4

vitvakatu commented:
Status update:
1. This bug is caused by implementation details of the wide component group, each column reports a lot of unnecessary `selected_entry` events, and some columns report `None` on every mouse move, making `on_change()` combinator effectively useless. I'm refactoring FRP to fix this issue. I expect to make a fix in the nearest hours.
2. This issue is unrelated to the PR itself, it's a bug in list view implementation. It's pretty hard to reproduce it reliably, I've created a new pivotal issue for that: https://www.pivotaltracker.com/story/show/182358403
3. Selecting non-visible entries below/above visible (non-clipped by scroll area) part of the component group would be fixed in a PR for task #182194574 (selection box).

<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/182358403">List View reports non-existing entry as selected - Pivotal Tracker</a></strong></div></blockquote>

vitvakatu committed:
Silence duplicated `selected_entry` events from wide component group

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/multiview-182259906

vitvakatu commented:
Status report: After messing up with a complex solution I reconsidered and made a much simpler one. So now the `selected_entry` events are not get duplicated in wide component group:


https://user-images.githubusercontent.com/6566674/171829651-24f5d72b-4353-492f-9fc2-76d831388a54.mp4

vitvakatu commented:
Unit tests in the component group module are failing, investigating.

vitvakatu committed:
wip

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/multiview-182259906

vitvakatu committed:
Implement overlay shape for list view

vitvakatu committed:
Add is_mouse_over output everywhere and use it in multiview wrapper

vitvakatu committed:
Disable pointer events for list-view::selection shape

vitvakatu committed:
revert some experiments

vitvakatu committed:
polishing the code

vitvakatu committed:
Formatting

vitvakatu commented:
Status report: screencast updated, spotted bugs fixed. There are two bugs still in place:
1. [Reported previously](https://www.pivotaltracker.com/story/show/182358403), on certain screen resolutions you can select an invisible entry.
2. When moving the mouse too quickly you can select an invisible entry as well, the discussion is ongoing in internal discord.

vitvakatu committed:
Rename local variable

vitvakatu committed:
Fix header selection

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/multiview-182259906

wdanilo commented:
What happens to this PR? I accepted it 14 days ago (!). Why do we still work on it?

mergify[bot] committed:
Merge branch 'develop' into wip/vitvakatu/multiview-182259906
